cfg/opencv2.cfg

```diff
@@ -22,11 +34,54 @@
   <define name="CV_IS_SUBMAT(flags)" value="((flags) &amp; CV_MAT_SUBMAT_FLAG)"/>
   <define name="MIN(a,b)" value="((a) &gt; (b) ? (b) : (a))"/>
   <define name="MAX(a,b)" value="((a) &lt; (b) ? (b) : (a))"/>
+  <define name="CV_Error(code,msg)" value="cv::error( code, msg, CV_Func, __FILE__, __LINE__ )"/>
+  <define name="CV_Assert(expr)" value="do { if(!!(expr)) ; else cv::error( cv::Error::StsAssert, #expr, CV_Func, __FILE__, __LINE__ ); } while(0)"/>
   <!-- ########## OpenCV containers ########## -->
   <!-- OpenCV containers that are similar to std containers -->
   <container id="cvString" startPattern="cv :: String" inherits="stdString"/>
   <!-- ########## OpenCV Allocation / Deallocation ########## -->
+  <memory>
+    <alloc init="false" buffer-size="malloc">cv::fastMalloc</alloc>
+    <dealloc>cv::fastFree</dealloc>
+  </memory>
   <!-- ########## OpenCV Functions ########## -->
+  <!-- void cv::errorNoReturn	(int _code, const String & _err, const char * _func, const char * _file, int _line) -->
+  <function name="cv::errorNoReturn">
+    <noreturn>true</noreturn>
+    <arg nr="1" direction="in">
+      <not-uninit/>
+    </arg>
+    <arg nr="2" direction="in">
+      <not-uninit/>
+    </arg>
+    <arg nr="3" direction="in">
+      <not-uninit/>
+    </arg>
+    <arg nr="4" direction="in">
+      <not-uninit/>
+    </arg>
+    <arg nr="5" direction="in">
+      <not-uninit/>
+    </arg>
+  </function>
+  <!-- void cv::fastFree (void *ptr) -->
+  <function name="cv::fastFree">
+    <noreturn>false</noreturn>
+    <returnValue type="void"/>
+    <arg nr="1">
+      <not-uninit indirect="1"/>
```

Comment:
@danmar Is the `indirect="1"` correct for a `free` like argument? I wonder because it is not yet specified this way for `free()`, but maybe `indirect` is just too new.

Reply:
I have the feeling that indirect should be 0 then. the data can be uninitialized but the pointer must be initialized.

Reply:
From the documentation:

> The indirect attribute can be set to control the indirection of uninitialized memory allowed. Setting indirect to 0 means no uninitialized memory is allowed. Setting it to 1 allows a pointer to uninitialized memory. Setting it to 2 allows a pointer to pointer to uninitialized memory.

> By default, cppcheck will use an indirect value of 0 unless not-null is used. When not-null is used, then indirect will default to 1.

I agree that the pointer must be initialized but not the memory that it points to.
As far as I understand the documentation the value 1 means that the pointer must be initialized and the memory can be uninitialized.
And 0 (the default) seems to mean that everything must be initialized, all pointers and the memory they point to.

Reply:
aha. I think you are right.

Reply:
Something looks wrong to me then.
For example with this config:
```xml
<?xml version="1.0"?>
<def format="2">
  <function name="foo">
    <arg nr="1">
      <not-uninit/>
    </arg>
  </function>
</def>
```
And this C file:
```c
#include <stdlib.h>
void f()
{
    char * p = (char *)malloc(20);
    char buf[20];
    foo(p);
    foo(buf);
}
```
I get no warning about uninitialized memory, but the default (0) for `indirect` should warn.
What is strange to me is that when I add `<not-null/>` to the function configuration, then I get errors for both calls of `foo`. But since `indirect` is set to 1 in this case (according to the documentation and comments), I would no longer expect errors.
Related PR: #2344

Reply:
I have created a ticket for the regression: https://trac.cppcheck.net/ticket/9705

Reply:
sorry for the late reply.. so the manual should be updated.

It should not claim that the default behavior when `indirect` is missing, is to check both pointer and data.

I want that our old configurations work as before. If you do not specify `indirect` then the configuration will match the old configurations.

Old behavior:

If you provide `<not-null/>` then the pointer must be initialized.

If you provide `<not-uninit/>` that gives you 1 additional level in the checking:
 * if `<not-null/>` is provided also, Cppcheck will also check that the pointer data is initialized.
 * otherwise, Cppcheck will just check that the pointer is uninitialized.

Reply:
I am taking a look at the manual..

Reply:
hmm.. I can't get it to work just with `indirect`

Reply:
hmm.. the documentation is really confusing..

It seems that the meaning of `<not-uninit/>` has changed over the years. :-(

If the manual has the original meaning (everything must be initialized) that sounds pretty intuitive ... then I wonder why and when that meaning was changed.

Reply:
Just adding a link to the documentation: https://github.com/danmar/cppcheck/blob/master/man/reference-cfg-format.md

Reply:
I updated the manual as good as I could.. please feel free to take a look. Would it be possible to clarify it better?

The configuration with `indirect` does not seem to work as I expect.

Reply:
@pfultz2 could you review my changes? 

296a0adc97665b1e3ae3938a0702ddc8e5eb620c

The `indirect` does not work as far as I see.. have I broken it or something?

Reply:
> The indirect does not work as far as I see.. have I broken/misunderstood it?

With the configuration in the example, Cppcheck does not warn.

Reply:
@pfultz2 Just a reminder: What do you think about the discussion above?

@danmar Do you think I can merge this PR? As far as I see the `indirect="1"` is not wrong.

Reply:
I think there is something wrong with the logic. Here we check that indirect != 0:

https://github.com/danmar/cppcheck/blob/master/lib/checkuninitvar.cpp#L1362

But later we check that indirect == 1:

https://github.com/danmar/cppcheck/blob/master/lib/checkuninitvar.cpp#L1362

Which should always be false.

Reply:
versat please feel free to merge this configuration... we will try to fix Cppcheck so it works as it should..

Reply:
> I think there is something wrong with the logic. Here we check that indirect != 0:
> 
> https://github.com/danmar/cppcheck/blob/master/lib/checkuninitvar.cpp#L1362
> 
> But later we check that indirect == 1:
> 
> https://github.com/danmar/cppcheck/blob/master/lib/checkuninitvar.cpp#L1362
> 
> Which should always be false.

I also wondered about this. But I am not sure how to best fix it.

Reply:
@versat in the code we have this:

            const bool isuninitbad = mSettings->library.isuninitargbad(start->previous(), argumentNumber + 1, indirect);
            if (alloc != NO_ALLOC)
                return isnullbad && isuninitbad;

so it seems you still need to configure `<not-null/>` also.

Reply:
hmm. I try to change it..